packages/flame/lib/src/components/component.dart

```diff
@@ -56,30 +74,135 @@ abstract class Component {
+  /// the component. Nevertheless, this method only need to be waited to finish
+  /// if by some reason, your logic needs to be sure that the component has
+  /// finished loading, otherwise, this method can be called without waiting
+  /// for it to finish as the BaseGame already handle the loading of the
+  /// component.
+  ///
+  /// *Note:* Do not add components on the game constructor. This method can
+  /// only be called after the game already has its layout set, this can be
+  /// verified by the [Game.hasLayout] property, to add components upon game
+  /// initialization, the [onLoad] method can be used instead.
+  Future<void> add(Component component) {
+    print('Added $component to $this');
+    return children.addChild(component);
+  }
+
+  /// Adds multiple children.
+  ///
+  /// See [add] for details.
+  Future<void> addAll(List<Component> components) {
+    return children.addChildren(components);
+  }
+
+  /// Removes a component from the component list, calling onRemove for it and
+  /// its children.
+  void remove(Component c) {
+    children.remove(c);
+  }
+
+  /// Removes all the children in the list and calls onRemove for all of them
+  /// and their children.
+  void removeAll(Iterable<Component> cs) {
+    children.removeAll(cs);
+  }
+
+  /// Whether the children list contains the given component.
+  ///
+  /// This method uses reference equality.
+  bool contains(Component c) => children.contains(c);
+
+  /// Call this if any of this component's children priorities have changed
+  /// at runtime.
+  ///
+  /// This will call `rebalanceAll` on the [children] ordered set.
+  void reorderChildren() => children.rebalanceAll();
+
+  /// This method first calls the passed handler on the leaves in the tree,
+  /// the children without any children of their own.
+  /// Then it continues through all other children. The propagation continues
+  /// until the handler returns false, which means "do not continue", or when
+  /// the handler has been called with all children
+  ///
+  /// This method is important to be used by the engine to propagate actions
+  /// like rendering, taps, etc, but you can call it yourself if you need to
+  /// apply an action to the whole component chain.
+  /// It will only consider components of type T in the hierarchy,
+  /// so use T = Component to target everything.
+  bool propagateToChildren<T extends Component>(
+    bool Function(T) handler,
+  ) {
+    var shouldContinue = true;
+    for (final child in children) {
+      if (child is BaseComponent) {
+        shouldContinue = child.propagateToChildren(handler);
+      }
+      if (shouldContinue && child is T) {
+        shouldContinue = handler(child);
+      } else if (shouldContinue && child is BaseGame) {
+        shouldContinue = child.propagateToChildren<T>(handler);
+      }
+      if (!shouldContinue) {
+        break;
+      }
+    }
+    return shouldContinue;
+  }
+
```

Comment:
/// Finds the closest parent up the hierarchy that satisfies type=T, or null if none is found